### PR TITLE
Bazel protobuf and gtest

### PR DESCRIPTION
**Pull request by wangkuiyi**
Fixes https://github.com/PaddlePaddle/Paddle/issues/816

In this PR, 

1. I initialize Bazel WORKSPACE file to declare an external dependency to https://github.com/pubref/rules_protobuf, which implements the Google-style protobuf building rules. 

   For more information please refer to this example repo: https://github.com/wangkuiyi/learn-bazel.

1. I also add a simple package in `/third_party/protobuf_test` to demonstrate using protobuf and gtest.

1. I updated the `docker_install_en.rst` documents to show how we can develop using Docker and Bazel.

----

Dear reviewers:

Please feel free to use Docker and Bazel to check this PR as described [in my update to Docker document in this PR](https://github.com/PaddlePaddle/Paddle/pull/818/files#diff-ceff150fa8464c5d2d9cc836209c65b5).

### CODE REVIEW

**Comment by wangkuiyi** [comment]
Following discussions in https://github.com/PaddlePaddle/Paddle/pull/819#pullrequestreview-12387533, I changed WORKSPACE to use Google protobuf as the external dependency for protobuf. @reyoung @lipeng-unisound Please review.

**Comment by reyoung** [comment]
The whole plan about bazel build, see #792.

@wangkuiyi We could edit the issue for every updates.

**Comment by wangkuiyi** [comment]
@reyoung Good point. To keep each PR a reasonablly small size, I splitted the work "incorporate glog, gtest, gflags, protobuf" listed in #792 into two:

1. incorporate gtest and protobuf #816 
1. incorporate glog and gflags #822

This PR fixes #816 . Since it incorporate gtest, someone else working on a PR fixing #822 can use gtest.

File: doc/getstarted/build_and_install/docker_install_en.rst [hunk 1]
```diff
@@ -104,3 +104,70 @@ container:
...
+
+
+3. Run the image as a container and mounting local source code
+   directory into the container.  This allows us to change the code on
+   the host and build it within the container.
+
+   .. code-block:: bash
+
+      docker run \
+       -d # run the container in background mode \
```
Comment (lipeng-unisound) [inline doc/getstarted/build_and_install/docker_install_en.rst]: The backslash continues the comments to next lines, thus makes the ```docker run``` command wrong.

**Pull request merged**